File: xdm-core/src/main/java/xdm/core/util/PlatformUtils.java
package xdm.core.util;

import xdm.core.constants.AppConstants;

import java.io.File;
import java.net.URISyntaxException;

public final class PlatformUtils {
  private PlatformUtils() {}

  public static String getConfigDir() {
    return new File(System.getProperty("user.home"), AppConstants.CONFIG_DIR).getAbsolutePath();
  }

  public static String getMetaDir() {
    return new File(
            System.getProperty("user.home"),
            AppConstants.CONFIG_DIR + File.separator + AppConstants.META_DIR)
        .getAbsolutePath();
  }

  public static String getExecutableFromSystemPath(String executable) {
    String path = System.getenv("PATH");
    if (!StringUtils.isNullOrEmpty(path)) {
      for (String spath : path.split(File.pathSeparator)) {
        File f = new File(spath, executable);
        if (f.exists()) {
          return f.getAbsolutePath();
        }
      }
    }
    return null;
  }

  public static boolean isWindows() {
    return System.getProperty("os.name").toLowerCase().contains("windows");
  }

  public static File getBaseDirectory() {
    try {
      File f =
          new File(
              PlatformUtils.class
                  .getProtectionDomain()
                  .getCodeSource()
                  .getLocation()
                  .toURI()
                  .getPath());
      return f.getParentFile();
    } catch (URISyntaxException e) {
      // No action
    }
    return null;
  }
}

File: xdm-core/src/main/java/xdm/core/util/PlatformUtils.kt
package xdm.core.util

import java.io.File
import java.net.URISyntaxException
import java.util.*

object PlatformUtils {
    fun getExecutableFromSystemPath(executable: String): String? {
        val path: String? = System.getenv("PATH")
        if (!path.isNullOrEmpty()) {
            for (p in path.split(File.pathSeparator)) {
                val f = File(p, executable)
                if (f.exists()) {
                    return f.absolutePath
                }
            }
        }
        return null
    }

    val isWindows: Boolean
        get() = System.getProperty("os.name").lowercase(Locale.getDefault()).contains("windows")

    val baseDirectory: File?
        get() {
            try {
                val f =
                    File(
                        PlatformUtils::class.java
                            .protectionDomain
                            .codeSource
                            .location
                            .toURI()
                            .path
                    )
                return f.parentFile
            } catch (e: URISyntaxException) {
                // No action
            }
            return null
        }
}
